### ISSUE TITLE
Server stop hook to avoid inconsistent state

### ISSUE THREAD

**Issue by ignopeverell** [mimblewimble/grin#536]
Add a hook on SIGINT to shutdown cleanly, especially when we're in the process of saving a block.

**Comment by yeastplume** [comment]
Let me introduce a complication.. I was actually thinking the exact same thing a while ago when doing some work on the mining thread. We don't see this day-to-day just yet, but when mining using the much slower cuckoo 30 on multiple GPUs, there can be a very noticeable delay between hitting ctrl-c and the underlying threads controlling CUDA operations etc finishing up, meaning there's the potential for a very long and very user-unfriendly delay there when trying to stop the server.

Whatever the solution is here, it needs to be some kind of service that different threads in different parts of the code can register to monitor, do their cleanup activities (and output whatever appropriate status messages if there's a delay) then only exit when all threads registered to listen for the handler have reported back.

**Comment by sesam** [comment]
Also in #502
There is https://doc.rust-lang.org/1.12.0/libc/constant.SIGHUP.html (libc wrapper) and apart from that rust is said to not support unix signals; https://users.rust-lang.org/t/unix-signals-in-rust/733

tl;dr use libc for now

**Comment by antiochp** [comment]
We probably want to differentiate between threads that can potentially write "corrupted" data and threads that can be safely (if not cleanly) shut down.
And yeah - rust has some big limitations in how it handles unix signals.

**Comment by heunglee** [comment]
Since I always encounter this type of issue with API servers, I read this issue here for better understanding of limitation on rust. I experienced some nodejs stuff for graceful shutdown. When googling, I did not have good luck to find good one for rust.  The rust book has a piece of graceful shutdown  which you probably already read and barely remember (https://doc.rust-lang.org/book/second-edition/ch20-06-graceful-shutdown-and-cleanup.html). LIke @sesam mentioned, if ctrl-c is not captured well, then the way of graceful shutdown won't be very useful. Yet it is worth of trying to feel out the roadblocks for graceful shutdown on multi threads?

**Comment by heunglee** [comment]
I have been some research on graceful shutdown for grin. 

For graceful shutdown, it may be better to experiment with the wallet listener first.
We can experiment with https://docs.rs/graceful/0.1.1/graceful/

Description of the crate "graceful" for those who have not had chance to get to know it before.
Gracefully shutdown.
(Unix) Async signals are tricky to handle properly. This utility let you block the main thread and execute arbitrary (thread-safe) code to shutdown the process gracefully. (I am not sure how much this crate is stable yet.)

The server side is much more complicated for graceful shutdown. So it is better to challenge the easy one, "wallet listener" first.

Currently the wallet listener starts wallet::server::start_rest_apis(wallet_config, keychain).
Before the Restful API server starts, coinbase and wallet receiver handlers are registered to the API server. The API server has stop function that can be called during graceful shutdown.

However, the phased tasks for aggsig transactions take place in wallet receiver. In the tasks, tx's are pushed to mempool via wallet client and wallet data are written. These tasks need to be waited to be finished during shutdown. As far as I know, the handler of received transactions in wallet receiver is not safe for killing wallet process. 

Since it has stop function , ApiServer can be stopped gracefully. It seems need a mechanism for wallet receiver to is notified to finish jobs during graceful shutdown. Maybe spawning child threads for each transaction for management during shutdown???

The steps might be 1. wait for wallet receiver handlers finishing and 2. stop ApiServer for graceful shutdown.

**Comment by heunglee** [comment]
I performed some experiments with wallet listener.
For graceful shutdown, I used tokio-signal (https://github.com/alexcrichton/tokio-signal). It works fine with SIGINT and SIGTERM on my Mac OS. (Not tried on linux yet)

Currently, when it receives SIGINT or SIGTERM, the grin thread calls ApiServer.stop() which just dereferences iron listening. There is an issue with iron listening.close() due to underlying hyper issue.
(fn close(&mut self) -> Result<(), Error>
Warning: This function doesn't work. The server remains listening after you called it. See https://github.com/hyperium/hyper/issues/338 for more details.) So the current way of shutdown cannot guarantee waiting for completion of processing coinbase and transfer transactions.

For complete graceful shutdown, it needs some mechanism to monitor status of transaction processing in handlers (WalletReceiver and CoinbaseHandler). In complicated way, using reactor??? Any suggestions?

Experimental Implementation:
// function called at wallet command : listening
fn handle_wallet_listening(wallet_config: WalletConfig, keychain: Keychain) {
	
	let srv = wallet::server::start_rest_apis(wallet_config, keychain);

	match srv {
		Some(mut apis) => {
			let mut core = Core::new().unwrap();
			let handle = core.handle();

			// Create a stream for each of the signals we'd like to handle.
			let sigint = Signal::new(SIGINT, &handle).flatten_stream();
			let sigterm = Signal::new(SIGTERM, &handle).flatten_stream();

			// Use the `select` combinator to merge these two streams into one
			let stream = sigint.select(sigterm);

			// Wait for a signal to arrive
			let (sgnl, _rest) = core.run(stream.into_future()).ok().unwrap();

			// Figure out which signal we received
			let sgnl = sgnl.unwrap();
			if sgnl == SIGINT || sgnl == SIGTERM {
				if sgnl == SIGINT {
					println!("SIGINT received.");
				} else {
					assert_eq!(sgnl, SIGTERM);
					println!("SIGTERM received");
				}
				info!(
					LOGGER,
					"Shutting down the Grin wallet listener...",
				);
				apis.stop();
				info!(
					LOGGER,
					"The Grin wallet listener has stopped.",
				);
			}

		},
		None => println!("Please try again after the issue being resolved."),
	}

}

**Comment by kargakis** [comment]
Seems like the issue is fixed for hyper 0.11 and there is an issue to get iron updated: https://github.com/iron/iron/issues/357

**Comment by heunglee** [comment]
I checked iron and hyper. As @kargakis mentioned, hyper 0.11 has run_until which allows graceful shutdown with future and reactor internally. However, iron still depends on hyper 0.10. So for now, an option for wallet is give some time interval before cleaning up iron listener so that handlers can get some time to finish the current tasks, which is not very safe in data protection though.

Another option is switching over to hyper 0.11. In this option, we lose path routing feature of iron. However grin has few routers, It can be done manually with hyper. But this will be a breaking change.

The best solution is to use hyper 0.11 for wallet listener in my opinion. Since there is no announced schedule of iron upgrade to hyper 0.11, we may need wait for some undefined time.

I will look into the server part for graceful shutdown.

**Comment by heunglee** [comment]
Re: miner plug-in.

Please refer to the following conversations below.
Heung B Lee @heunglee 13:08
@yeastplume Question about miner plugin. What would be an graceful way to stop the plug-in? Using thread join? It works fine with linux, but mac seems have some issue per rust doc.

Yeastplume @yeastplume 13:09
@heunglee that's a good question.. the problem is that the ability to interrupt the solve needs to be built into the solver itself, which it currently isn't
that particular thread won't return until CUDA's done it's thing.. I tried to implement something a while ago to interrupt it, but it caused further problems, and I don't think it'll be possible to do without inserting checks that could impact on solve times

Heung B Lee @heunglee 13:12
So that means, can stop the miner on returning once CUDA's done an iteration?

Yeastplume @yeastplume 13:13
Yes, that's easy enough to do. You can stop it gracefully (there's a stop function in cuckoo-miner for async mode), or just don't call the mine function again if you're in sync mode, but you'll usually have to wait for the current solve to finish

Heung B Lee @heunglee 13:17
Thank you. I am going to put these conversations in a related issue for reference.

**Comment by yeastplume** [comment]
I believe we have this now
